Follow API (2)

akshays2112 opened the pull request:
Hi Daniel,

I hope I fixed everything. Going to go one at a time serially. Here are FollowApi changes.

Thanks for your time and patience.

Best Regards,

Akshay Srinivasan

akshays2112 committed:
FollowApi updated and working.

akshays2112 committed:
Changed in FollowApiTests back to var.

akshays2112 committed:
Checked SpotifyApi documentation and corrected the summary section in FollowApi and IFollowApi.

DanielLarsenNZ commented:
OK thanks, taking a look now

DanielLarsenNZ commented on src/SpotifyApi.NetCore/SpotifyWebApi.cs:
```diff
@@ -173,10 +179,13 @@ private async Task<HttpResponseMessage> PostOrPut(string verb, Uri uri, object d
             {
                 case "PUT":
-                    response = await _http.PutAsync(uri, content);
+                    response = await _http.PutAsync(uri, data == null ? null : content);
```
Don't change this behaviour - `content` is the correct argument value. Your refactoring tool may be adding unnecessary null checks here?

akshays2112 replied:
Hi Daniel,

Experimenting I found that my code for PUT it has to be this way or it fails with:

 FollowArtists_ArtistIds_IsTrue
   Source: FollowApiTests.cs line 75
   Duration: 200 ms

  Message: 
    Test method SpotifyApi.NetCore.Tests.FollowApiTests.FollowArtists_ArtistIds_IsTrue threw exception: 
    SpotifyApi.NetCore.SpotifyApiErrorException: Invalid JSON body
  Stack Trace: 
    RestHttpClient.CheckForErrors(HttpResponseMessage response) line 119
    SpotifyWebApi.PostOrPutOrDelete(String verb, Uri uri, Object data, String accessToken) line 196
    SpotifyWebApi.Put(Uri uri, Object data, String accessToken) line 154
    FollowApi.FollowArtists(String[] artistIds, String accessToken) line 162
    FollowApiTests.FollowArtists_ArtistIds_IsTrue() line 85

If you can correct my code and explain what I am doing wrong then would be nice. Otherwise I have to give up on doing all PUT's and POST's because it does not work as it is.

Thanks for your time and patience.

Best Regards,

Akshay Srinivasan

DanielLarsenNZ replied:
OK I will take a look

DanielLarsenNZ replied:
Hmmn, it looks like this is handled inconsistently by the Spotify API. Leave it as it is for now @akshays2112 and I will do some more investigation and fix it after the merge 👍

DanielLarsenNZ commented on src/SpotifyApi.NetCore/SpotifyWebApi.cs:
```diff
@@ -173,10 +179,13 @@ private async Task<HttpResponseMessage> PostOrPut(string verb, Uri uri, object d
                     break;
                 case "POST":
-                    response = await _http.PostAsync(uri, content);
+                    response = await _http.PostAsync(uri, data == null ? null : content);
```
Don't change this. The null check is not correct.

DanielLarsenNZ commented on src/SpotifyApi.NetCore/Models/PagedArtists.cs:
```diff
@@ -0,0 +1,13 @@
+    /// Full Artist Object.
+    /// </summary>
+    /// <remarks> https://developer.spotify.com/documentation/web-api/reference/object-model/ </remarks>
+    public class PagedArtists : Paged<Artist>
```
Nice one 😎

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+
+        #region UnfollowPlaylist
+        /// <summary>
+        /// Unfollow a Playlist
```
This should be "Remove the current user as a follower of a playlist." as per docs

akshays2112 replied:
Hi Daniel,

Here you are not going by the title in the docs you are going by the full description.  I am going by the title everywhere.  Need to get a decision on the title or full sentence description for this and I change it once everywhere.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

DanielLarsenNZ replied:
That's right. The title is used for the Method name, i.e. "Follow a Playlist" = `FollowPlaylist`

The description is used for the `<summary>` tag, i.e.

![image](https://user-images.githubusercontent.com/2294760/85985240-8b896f00-ba3e-11ea-96ba-c263e9dc0502.png)

akshays2112 replied:
Hehe did not need the image though *giggles* but I am happy now I was very depressed that I could not pass this code. ♥♥♥

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+            );
+
+        /// <summary>
+        /// Unfollow Artists or Users
```
This should be "Remove the current user as a follower of one or more other Spotify users." as per docs

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+
+        #region UnfollowArtistsOrUsers
+        /// <summary>
+        /// Unfollow Artists or Users
```
This should be "Remove the current user as a follower of one or more artists." as per docs

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+            );
+
+        /// <summary>
+        /// Get User's Followed Artists
```
"Get the current user’s followed artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+
+        #region GetUsersFollowedArtists
+        /// <summary>
+        /// Get User's Followed Artists
```
"Get the current user’s followed artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+
+        #region FollowPlaylist
+        /// <summary>
+        /// Follow a Playlist
```
"Add the current user as a follower of a playlist."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+            );
+
+        /// <summary>
+        /// Follow Artists or Users
```
"Add the current user as a follower of one or more other Spotify users."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -93,5 +93,119 @@ public interface IFollowApi
+
+        #region FollowArtistsOrUsers
+        /// <summary>
+        /// Follow Artists or Users
```
"Add the current user as a follower of one or more artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -79,10 +79,10 @@ public interface IFollowApi
 
         /// <summary>
-        /// Check to see if one or more Spotify users are following a specified playlist.
+        /// Check if Users Follow a Playlist
```
"Check to see if one or more Spotify users are following a specified playlist."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -1,73 +1,73 @@
         #region CheckCurrentUserFollowsPlaylist
         /// <summary>
-        /// Check to see if one or more Spotify users are following a specified playlist.
+        /// Check if Users Follow a Playlist
```
"Check to see if one or more Spotify users are following a specified playlist."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -1,73 +1,73 @@
 
         /// <summary>
-        /// Check if Current User Follows Users
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more other Spotify users."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -1,73 +1,73 @@
 
         /// <summary>
-        /// Check if Current User Follows Users
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more other Spotify users.

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -1,73 +1,73 @@
 
         /// <summary>
-        /// Check if Current User Follows Artists
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IFollowApi.cs:
```diff
@@ -1,73 +1,73 @@
+        #region CheckCurrentUserFollowsArtistsOrUsers
         /// <summary>
-        /// Check if Current User Follows Artists
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -11,102 +12,97 @@ namespace SpotifyApi.NetCore
-
         /// <summary>
-        /// Check if Current User Follows Artists
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -11,102 +12,97 @@ namespace SpotifyApi.NetCore
 
         /// <summary>
-        /// Check if Current User Follows Artists
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -11,102 +12,97 @@ namespace SpotifyApi.NetCore
+                    ArgumentException("A minimum of 1 and a maximum of 50 artist ids can be sent.");
 
-            var builder = new UriBuilder($"{BaseUrl}/me/following/contains");
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following/contains");
```
Please use `var` here

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -11,102 +12,97 @@ namespace SpotifyApi.NetCore
+                    ArgumentException("A minimum of 1 and a maximum of 50 user ids can be sent.");
 
-            var builder = new UriBuilder($"{BaseUrl}/me/following/contains");
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following/contains");
```
Use `var` here. Your refactoring tooling maybe making these changes?

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -11,102 +12,97 @@ namespace SpotifyApi.NetCore
 
         /// <summary>
-        /// Check if Current User Follows Users
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more other Spotify users."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -11,102 +12,97 @@ namespace SpotifyApi.NetCore
 
         /// <summary>
-        /// Check if Current User Follows Users
+        /// Check if Current User Follows Artists or Users
```
"Check to see if the current user is following one or more other Spotify users."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -11,102 +12,97 @@ namespace SpotifyApi.NetCore
         #region CheckCurrentUserFollowsPlaylist
         /// <summary>
-        /// Check to see if one or more Spotify users are following a specified playlist.
+        /// Check if Users Follow a Playlist
```
"Check to see if one or more Spotify users are following a specified playlist."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -118,10 +114,10 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
 
         /// <summary>
-        /// Check to see if one or more Spotify users are following a specified playlist.
+        /// Check if Users Follow a Playlist
```
"Check to see if one or more Spotify users are following a specified playlist."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+                    ArgumentException("A minimum of 1 and a maximum of 5 user ids can be sent.");
 
-            var builder = new UriBuilder($"{BaseUrl}/playlists/{playlistId}/followers/contains");
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/playlists/{playlistId}/followers/contains");
```
Use `var` here

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+
+        #region FollowArtistsOrUsers
+        /// <summary>
+        /// Follow Artists or Users
```
"Add the current user as a follower of one or more artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            if (artistIds?.Length < 1 || artistIds?.Length > 50) throw new
+                    ArgumentException("A minimum of 1 and a maximum of 50 artist ids can be sent.");
+
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following");
```
use `var` here

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+        }
+
+        /// <summary>
+        /// Follow Artists or Users
```
"Add the current user as a follower of one or more other Spotify users."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            if (userIds?.Length < 1 || userIds.Length > 50) throw new
+                    ArgumentException("A minimum of 1 and a maximum of 50 user ids can be sent.");
+
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following");
```
Use `var` here

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+
+        #region FollowPlaylist
+        /// <summary>
+        /// Follow a Playlist
```
"Add the current user as a follower of a playlist."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            if (string.IsNullOrWhiteSpace(playlistId)) throw new
+                    ArgumentNullException("playlistId");
+
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/playlists/{playlistId}/followers");
```
Use `var` here

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+
+        #region GetUsersFollowedArtists
+        /// <summary>
+        /// Get User's Followed Artists
```
"Get the current user’s followed artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            ) => await GetUsersFollowedArtists<PagedArtists>(limit, after, accessToken);
+
+        /// <summary>
+        /// Get User's Followed Artists
```
"Get the current user’s followed artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            if (limit < 1 || limit > 50) throw new
+                    ArgumentException("A minimum of 1 and a maximum of 50 artist ids can be sent.");
+
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following");
```
Use `var` here

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following");
+            builder.AppendToQuery("type", "artist");
+            builder.AppendToQuery("limit", limit);
+            if (!string.IsNullOrWhiteSpace(after))
+            {
+                builder.AppendToQuery("after", after);
+            }
```
There is an extension method for this:

    builder.AppendToQueryIfValueNotNullOrWhiteSpace("after", after);

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+
+        #region UnfollowArtistsOrUsers
+        /// <summary>
+        /// Unfollow Artists or Users
```
"Remove the current user as a follower of one or more artists."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            if (artistIds?.Length < 1 || artistIds?.Length > 50) throw new
+                    ArgumentException("A minimum of 1 and a maximum of 50 artist ids can be sent.");
+
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following");
```
Use `var`

akshays2112 replied:
Hi Daniel,

Will change this everywhere.  If I understand the logic you will use var when the words match meaning in this case since UriBuilder string matches == UriBuilder there is no point typing it again so use var. Is that it? Because technically in my code there is no statement which will not resolve to a type in assignment operations hence technically all should be var.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

DanielLarsenNZ replied:
That is correct, use `var` wherever the Type is obvious.

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+        }
+
+        /// <summary>
+        /// Unfollow Artists or Users
```
"Remove the current user as a follower of one or more other Spotify users."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            if (userIds?.Length < 1 || userIds?.Length > 50) throw new
+                    ArgumentException("A minimum of 1 and a maximum of 50 user ids can be sent.");
+
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/me/following");
```
use `var`

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+
+        #region UnfollowPlaylist
+        /// <summary>
+        /// Unfollow a Playlist
```
"Remove the current user as a follower of a playlist."

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -132,14 +128,193 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
+            if (string.IsNullOrWhiteSpace(playlistId)) throw new
+                    ArgumentNullException("playlistId");
+
+            UriBuilder builder = new UriBuilder($"{BaseUrl}/playlists/{playlistId}/followers");
```
use `var`

DanielLarsenNZ commented:
Hey @akshays2112 - some good work here but I think your refactoring tooling may be making some unnecessary changes - are you using Resharper? If so, I suggest you customise your ruleset or turn it off for this project 🙂

You have also undone a lot of the summary comments that were correct. It is important that the XML comments match the Spotify docs so that users have a consistent experience 😃

Be really careful in `SpotifyWebApi` class - this is used by all API classes, so small changes can have a major impact. Great that you added the `Delete` method! Just watch out for incorrect `null` checks and just ask if you are not sure 😎👍 Thank you 🤩

akshays2112 commented:
Hi Daniel,

I have left some comments on specific types of issues. Will defer changes until we can come to an agreement to shorten the process.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

akshays2112 commented:
Hi Daniel,

YAY!!! Thanks so much for those replies I will work on those two types of issues that have been clarified.  Please look at PUT/POST or my code to see why my code is failing unless I do the null check.

Btw what I found out long ago that I need to do the null check because it is passing some unnecessary header which tells the parser to expect some JSON when there is none so obviously it fails.

**I figured out how to get it to work with the existing PUT and POST so I will modify all my code and remove my changes to SpotifyWebApi class except of course for DELETE which I added.**

♥♥♥ I was losing my mind here LOL.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

DanielLarsenNZ commented:
No worries @akshays2112 and sorry for the confusion 😎

akshays2112 committed:
Fixed all issues to do with SpotifyAPI descriptions in summary, var usage.
Got a fix for the problem I was facing with need null checks for PUT and everything is working.
Rolled back changes I made to SpotifyWebApi for the null checks.

akshays2112 committed:
Added some comments for my Helper Class code to assist in correctly working with existing PUT and POST requests without needing to change them.

akshays2112 committed:
Missed one small change.

akshays2112 commented:
Hi Daniel,

Any chance of closing this moving on to the next API I already coded.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

DanielLarsenNZ commented:
Sure thing @akshays2112 , I will take a look tonight :)

DanielLarsenNZ committed:
Use dynamic object for artist/user id array

DanielLarsenNZ commented:
Merged! Passing a body to the Put method instead of null is really smart @akshays2112 . This only minor change I have made is to use a dynamic object for the user/artist id array instead of your helper.

Whoop Follow API is done! 🥳🎉

akshays2112 commented:
Big boss thanks now I am going to do up LibraryApi and startup another pull request.

DanielLarsenNZ commented:
Sweet! Let's do this. I am going to bed now, but will take a look in the morning. Thanks again Akshay
